### ISSUE TITLE
Enhancement Request: X-Road Metrics version 2.x

### ISSUE THREAD

**Issue by melbeltagy** [nordic-institute/X-Road-Metrics#155]
## TL;DR
Current implementation of X-Road Metrics, although working properly and performing very well, has many pain-points from administration, security, and development point-of-views.
Proposed solution is to migrate existing applications from CLI-based applications into microservices architecture with AdminUI and UserUI components.

### Background information
Current implementation of X-Road Metrics, although working properly and performing very well, has the following pain-points from administration, security, and development point-of-views.

#### Administration point-of-view
 - Administrative tasks are scattered across different modules:
   - `Collector` module has scripts to create MongoDB's users, collections, and indexes
   - `OpenData` module has script to create PostegeSQL' users
   - While `Anonymizer` module creates the PostegeSQL' tables
 - For every module, maintaining the configurations requires physical access to the file system of the module's server and manually changing them
 - For every module, there are different operations that can be run. e.g., `Collector` has `list`, `collect`, and `update`, while `Anonymizer` has only one. The issue is, these operations are not clear to the administrator. 
 - Usually, only one operation is scheduled as a cron job on the server
 - For administrators to run any of these jobs/operations outside of their schedule, or to run an operation that does not have a schedule, once again, they have to connect to the running server's shell and execute the required operation(s) manually.
- Although applicable to build a docker image out of each module, running them in Kubernetes and maintaining the configurations is not an easy administrative task at all.

#### Security point-of-view
- Since each module's job is configured as a native cron job, the docker images will always require root privilege making the docker images not adhering to images best practices from security perspective (similar to the Security Server Sidecar image).
This issue actually prevents some of the Kubernetes-based runtime, e.g., OpenShift, from running the images without workarounds that some organizations consider a big risk.
- Storing configurations on the file-system, which contains database credentials is a big issue.
- Enforcing backwards compatibility with Python 3.8 and Ubuntu 20.4 (Focal) by preventing dependencies upgrade to latest versions.

#### Development point-of-view
- Each module has its own technical stack. Some are pure python CLI applications, and others are using Django framework. Then we have Networking module which is based on `R` language. Maintaining these different technical stacks in one repository is difficult affecting the cost of maintenance and adding new features.
- A lot of workaround has been in place to keep backwards compatibility with Python 3.8 and Ubuntu 20.4 (Focal).
- Despite the latest efforts of enhancing the developers experience, it's not easy to run the applications locally (must build and run in docker locally, which must be built by the developer since the repo does not have any mean of building images).
- For a developer to run any module, it requires administration tasks as well, exposing the developer with the same issues as an administrator
- The structure of each module is not according to the standard. e.g., tests are inside each package
- Each module has some utility command lines that are either obsolete or only handful of people are aware of how to use them or what they do
- The way code is structured, makes IDEs give errors and cannot work with the different imports and type checks (I tried many times to fix, but cannot without drastic changes)
- There are some modules that are not published yet, but part of the code base. Makes it hard to understand which parts to maintain and which to keep aside for now.

> [!IMPORTANT]
> Although each of the previous pain points can be addressed individually, it's really hard and time consuming given the current state. (I tried 😞 )

### Metrics Version 2.0

The proposed solution is:
* Migrate existing applications (modules) from being CLI applications to be microservices based on FastAPI framework. i.e., switching from CLI to REST APIs. This applies to all modules including `Networking`.
* Introduce Admin UI. A unified web application targeting administrators. It allows them to:
  - Run different operations/jobs supported by any of the `Collector`, `Corrector`, `Reports`, and `Anonymizer` applications.
  - Configure each application (including `OpenData` and `Networking`)
  - Download PDF reports from UI (instead of accessing the server's file system or waiting for emails)
    - _Emails would still be available_
* Introduce User UI. Another unified web application that is targeting public users (replacing existing Django and R UIs)
  - This way, Only `UserUI` and its proxy is in the DMZ zone while `OpenData` and `Networking` deployed in the internal network.
* Introduce a new module that would be responsible for all DB administrative tasks (creating MongoDB users, collections, and indexes. Migrate PostgreSQL' database using database migration tools. e.g., [Alembic](https://alembic.sqlalchemy.org/en/latest/)
* Default packaging would be Docker images.
  _Notes:_
  Sharing my thoughts about this point:
  - My preferred option is to actually drop support for Debian packaging.
  - If docker images are not the only required option, then the 2nd best option is to package the application as a python application and publish them to Artifactory. This way, we can support any OS and be more flexible.
  - If Debian packages are still needed, it can still be created and published.

The proposed solution will inherently provide the following fixes as well 
* Migrate all configurations to be environment variables or MongoDB (depending on the configuration's nature)
* Using Docker images means that we can always use latest Python and dependencies and hence security fixes.
* Docker images will adhere to best practices
* Fix code structure
* Cleanup the code base from unused code, libraries, and tools

#### Additional items to include
* Move extra modules that are not published yet to a different branch from `main`. i.e., Move `analysis_module`, `analysis_ui_module`, `archive_module`, `opendata_collector_module` to a `incubation` branch for enhanced clarity and maintainability.

**Comment by melbeltagy** [comment]
# Milestone 1

Regarding **Milestone 1**, I tried multiple approaches and found out that the simplest one is to move the non-released modules into a nested folder and disable any files that would cause Github or dependabot to consider the folders as python applications.  
Already implemented in PR #156  
_I personally highly recommend merging this PR as it will remove confusing code from developers' concern.  
They can be, of course, moved out of the folder when needed after being upgraded and fixed._

---

## Alternate Options

### Option 1

<ins>**Conclusion:**</ins> Causes a lot of hidden irreversible issues that are usually not noticed until it's too late (according to official doc and articles online)

Use [git filter-branch](https://git-scm.com/docs/git-filter-branch). Sounded very promising until read the warnings.  
I tried playing around with it, git command refused to execute it and pointed out to the second option below.

### Option 2

<ins>**Conclusion:**</ins> A big no go in our use-case.

The previous option points to [git-filter-repo](https://github.com/newren/git-filter-repo) as a replacement.  
In summary it's a tool that can be used to remove any traces of unwanted files or folders. It works great, but does not fit our use-case as it filters out contents on repo level, not branch level.

### Options 3

<ins>**Conclusion:**</ins> if really needed, can be done, but has drawbacks and seems to be overkill in our case.

Move each of non-released modules into its own incubation branch using [git subtree](https://www.atlassian.com/git/tutorials/git-subtree) command.  
This tool fits best in scenarios where a folder that represents a service or a library is required to be split into its own repository. It still works for our use-case but with caveats.

**Comment by VitaliStupin** [comment]
If a proper rewrite is considered for X-Road Metrics, performance and optimization issues should be addressed.

The application currently fulfills its purpose but scales poorly as the volume of X-Road queries increases. For example, three months of Estonian X-tee data requires 1TB of storage, with half occupied by indexes. Some commonly used indexes reach 40GB in size, and multiple indexes are utilized concurrently. While MongoDB is optimized for in-memory operations, it's infeasible to allocate hundreds of gigabytes of memory exclusively for MongoDB. A machine with 64GB of RAM struggles to process new metrics due to excessive I/O operations, as it cannot hold all required indexes in memory.

MongoDB might not be the ideal database for X-Road Metrics. However, when considering alternatives, it's crucial to acknowledge that any other database optimized for fast ingestion and less frequent updates would likely face similar challenges. A typical sequence of database operations to ingest metrics data into MongoDB is as follows:

* Collector: Adds PartyA (client or producer) data to the raw_messages collection.
* Corrector: Reads PartyA data from raw_messages.
* Corrector: Searches the xRequestId index to check if PartyB data has already been added to clean_data (in this flow, PartyB data is not yet available).
* Corrector: Adds PartyA data to the clean_data collection.
* Corrector: Marks PartyA data as processed in raw_messages.
* Collector: Adds PartyB data to raw_messages.
* Corrector: Reads PartyB data from raw_messages.
* Corrector: Searches the xRequestId index to check if PartyA data has already been added to clean_data.
* Corrector: Reads PartyA data from clean_data.
* Corrector: Updates PartyA data with fields from PartyB and saves it to clean_data.
* Corrector: Marks PartyB data as processed in raw_messages.
* Cleanup: Finds and deletes processed documents from raw_messages.
* Cleanup: Finds and deletes documents from clean_data older than X days (e.g., 90 days). This step is particularly slow as it requires checking and processing individual rows.

Some steps in this process are redundant. For example, processed documents in raw_messages could be directly deleted. Alternatively, it could be considered to bypass the raw_messages collection altogether and directly passing data from the collector to the corrector. This would slightly slow down the collection process but significantly reduce overall disk I/O.

Deleting old documents from the database is a very slow operation. Deleting a single day's worth of data can take hours. The primary issue is that MongoDB does not support table partitioning. This prevents the deletion of an entire month/week/day instantly, necessitating slow search/delete operations within MongoDB.

The current data model is also suboptimal. Client and producer data are separated into different sub-elements, and one of these sub-elements may be missing. This requires up to four separate searches for many operations, especially during report generation, instead of a single search. Furthermore, storing duplicate data in both client and producer elements introduces overhead. A more efficient approach would involve:

* Fields common to both client and producer.
* Aggregated fields for optimized searching and simplified reporting/visualizations.
* Client-specific fields without duplicates.
* Producer-specific fields without duplicates.

This structure would minimize the number of searches required and eliminate unnecessary conditions and checks for the existence of client or producer elements.

Report generation also necessitates optimization. Due to the inability of indexes to fit into memory, generating reports for members with numerous queries can be as slow as a full collection scan. Performing a single full scan and aggregating data for each member instead should be considered. Database support for partitioning would also significantly improve performance by reducing index sizes and eliminating the need to process data from previous months.

Currently, X-Road Metrics does not provide Kibana/Grafana visualizations and searches. The Estonian X-Road center synchronizes data from MongoDB to Elasticsearch and utilizes Kibana for problem identification and visualization of X-Road usage statistics. If major changes are implemented, the synchronization process may need to be adjusted. If the database is replaced, a new synchronization solution would be required. However, this should not be considered a major obstacle, as optimizing X-Road Metrics can significantly reduce hardware resource consumption.

**Comment by melbeltagy** [comment]
@petkivim , @raits , and @VitaliStupin, your help and input is needed, please.

I've created a Metrics 2.0 skeleton on my fork https://github.com/melbeltagy/X-Road-Metrics/tree/metrics-2-skeleton.
It's a branch without any history, so that it won't interfere with existing code-base, but this is a temporary step. 

So, I need your input reviewing the skeleton and discussing the ADRs included.

Running the skeleton is super easy, just clone the repo, checkout the branch, and run the docker compose.
All services will be built and run in Docker. No need to install any other tools unless updating the code is required.

To review the code and for now, recommended IDE is VS Code. Branch has some related configs with recommended extensions (once loaded by VS Code, recommended extensions alert will pop-up).

The branch includes a list of [Architecture Decision Records (ADRs)](https://github.com/melbeltagy/X-Road-Metrics/tree/metrics-2-skeleton/docs/adr). I marked some as approved either because they follow the same pattern used in Metrics v1.x, or because they were initially approved by accepting this RFE over slack.
Others are still `Draft`, meaning they require your approval please. Of course, feel free to revisit any ADR, even it was marked as `Accepted`.
As for the ADRs folder, the plan is to delete them later on. I just added them now as a way to communicate them and get approval first before moving forward.

### High-level intro about the code:
_For more info about the structure, please refer to [ADR-010](https://github.com/melbeltagy/X-Road-Metrics/blob/metrics-2-skeleton/docs/adr/ADR-010-monorepo-structure.md)_

#### Backend Services:

1. Have no business logic from Metrics 1.x **yet**
2. Each has `/health`, `/ready`, and `/info` [probes](https://kubernetes.io/docs/concepts/configuration/liveness-readiness-startup-probes/) for better k8s support
3. Each has `/greeting` endpoint that can be used to test the service and its logs.
4. Swagger is enabled by default in the docker compose file. Swagger links are in README.md file

The future plan is that admins, via `admin-ui` app, would update the configs, and services would retrieve configurations. Initial plan was to use MongoDB for storing configs, but with Vitali's input, I'm considering removing mongo DB entirely, but this is a future decision to be discussed and will need some kind of PoC first.
So, for now, seems that the safest option is to store configs in postgres instead. Based on this, all services now depends on postgres library, while only the ones that require mongo depends on mongo library. 

#### Libraries

Have some implementations. I tried to implement some libraries for me to understand and to illustrate how things are going to be connected. I hope this would give you enough info to make you decision-making process much easier and informed.

#### The apps (UI application):

1. Has no business implementation at all. 
2. Accessible, but based on Vue.js default template without any customization. 
3. Will use REST to call BE services .

**_note:_** 
At this stage, the valuable information that was provided by @VitaliStupin in this RFE are not taken into considerations yet (sorry).
The plan is as the following:
- Step 1: Move to Docker images with minimal change to the logic. Just focusing on Docker, FastAPI, admin UI, etc.
- Step 2: Refactor the processing logic to address Vitali's comment.
Taking it all into 1 shot is too much, hence the 2 steps so that it would be much easier to target and deliver.

Since you'll have A LOT to read and review in the branch, I'll stop now 😄 , but please feel free to ask any question or object on anything as you read/review.

**Comment by VitaliStupin** [comment]
I have read through the ADR documents; the decisions seem to be well-argued and viable. While I still have concerns regarding MongoDB, keeping it for now does make the initial rewrite easier. I am also not convinced that creating end user UI solely for report downloads is worth the effort, and using Keycloak just for admin access may be overkill.

Regarding the local setup, I encountered errors with the "just" command: "error: Non-default parameter args follows default parameter, justfile:89:27". I suspect a version mismatch, as the setup guide is quite minimal. For now, I have worked around this by commenting out the failing lines.

Additionally, I found that my current setup cannot run modern MongoDB containers because they require AVX support, which VirtualBox disables when the host is running Hyper-V. I managed to run docker-compose with Mongo v4.4, but depending on the features used, I may eventually need an external machine for the database.

The Git repository structure looks easy to work with, and the initial skeleton should allow developers to navigate the code easily. It is too early to tell if we will encounter the same multiprocessing issues as the previous codebase, so this should be monitored closely: https://github.com/nordic-institute/X-Road-Metrics/issues/177 .

Finally, any rewrite risks performance drops. We must test "corrector" and "reports" thoroughly once the rewrite is finished to ensure they remain performant, as we are currently close to the processing limits for the Estonian X-Road instance.

**Comment by melbeltagy** [comment]
Thank you so much @VitaliStupin for your quick and, as usual, valuable feedback. Much appreciated.

Regarding the UI part:
- Admin UI will be used to allow admins to configure X-Road Metrics (setup profiles and connection parameters to X-Road SS & CS, etc). Since reports are only accessible by the admins, the same Admin UI should also be used for that.
- Admin UI will require authentication. Since Metrics will be running in containers, I tried to avoid creating an OS-level user(s). Instead, the Admin UI will allow any OIDC based server to be used for authentication (as described by [ADR-007](https://github.com/melbeltagy/X-Road-Metrics/blob/metrics-2-skeleton/docs/adr/ADR-007-oidc-client-oidc-client-ts.md)). Keycloak is just one option and will be used by default for the development environment or new environments where an OIDC server is not available.
- Opendata-UI application will be used by public users. It should contain current functionality provided by existing `opendata_module` and `networking_module`. The name is not obvious and a better name could be used, but I couldn't come up with something that makes sense yet.

If we have a better suggestion to handle the UI apps, I'm all ears.

I was hoping that using `just` would be much smoother 😆 . Apologies that it didn't go as planned. Would you be able to share which part that failed, please?

And yes, MongoDB is just for initial rewrite. I'm inclined towards removing it entirely, but that's for the next step.

**Comment by VitaliStupin** [comment]
Hi!

```
$ just --version
just 1.21.0
$ just --list
error: Non-default parameter `args` follows default parameter
  ——▶ justfile:89:27
   │
89 │ docker action service="" *args:
   │    
```

Failing lines:
```
docker action service="" *args:
    docker compose -f docker/compose.yaml {{action}} {{service}} {{args}}
```

Regarding the UI, we are quite used to not having an Admin UI for configuration at all. This leads me to question whether Keycloak, or any other authentication solution, adds enough value right now, especially since we prefer to automate setup and avoid manual configuration changes. However, it is entirely possible that if we add functionality like custom reports in the future, an Admin UI would make much more sense.

**Comment by petkivim** [comment]
@melbeltagy Thank you very much for the effort you’ve put into X-Road Metrics! Please find below some comments - some of them are more detailed while other are more general.

- [ADR-012-secrets-management](https://github.com/melbeltagy/X-Road-Metrics/blob/metrics-2-skeleton/docs/adr/ADR-012-secrets-management.md) - The selected option is acceptable. However, if other options need to be considered, OpenBao would be the preferred choice. X-Road 8 will use OpenBao as its secret storage, which makes it a natural choice for X-Road Metrics as well.
- [ADR-015-deployment-options](https://github.com/melbeltagy/X-Road-Metrics/blob/metrics-2-skeleton/docs/adr/ADR-015-deployment-options.md) - This is just a minor detail regarding tagging the Docker images. The artifacts should follow the NIIS tagging strategy for Docker, e.g., the strategy doesn't include the `latest` tag. However, the strategy isn't publicly available or even applied in practice yet. It will be applied starting from the release of X-Road 8.


**About the UIs**

- Implementing an admin UI that allows configuring the Metrics instance is an important improvement and makes setting up and managing Metrics easier for new users. Nevertheless, fully automating the setup should be supported too.
- In addition to configuring the instance, it would be nice if the admin UI supports exploring the collected data too. Or alternatively, the default setup comes with a separate analytics and visualisation tool (e.g., Kibana, Grafana) that administrators can use to explore the collected data.  
- Initially, implementing a new UI for downloading reports can be optional since Metrics v1 doesn't have it either.
- The admin UI and the UI for downloading reports have different security boundaries, which is why they should be separated. The admin UI is accessed by X-Road Metrics administrators, and access should be allowed only from whitelisted IP addresses. In contrast, reports are downloaded by people representing X-Road member organisations, which means that the UI must be accessible from the Internet. For these reasons, the admin UI and the UI for downloading reports should be implemented as two separate applications so that access to them can be controlled independently.
- Using any OIDC-compliant solution as an identity provider sounds like a good approach. The default setup can use Keycloak as suggested, but replacing it with any other OIDC-compliant solution is also possible.

**About the architecture** 

When designing the architecture, it should be easily extensible. The aim is to expand Metric's functionality to cover environmental monitoring data too. Right now, there's X-Road Metrics for collecting operational monitoring data and a different solution ([X-Road Monitor Collector](https://github.com/vrk-kpa/xroad-monitor-collector) - it uses ElasticSearch for storing the collected data) to collect environmental monitoring data. Nevertheless, on a high level, both solution do exactly the same thing - they collect data from Security Servers, store it centrally and do some processing with it. Even if operational and environmental monitoring data sets are different, the overall process contains many same or similar elements. In the future, a third data set to be collected may be sustainability-related data unless it will be included in operational or environmental monitoring data sets. Instead of having three very similar solutions that collect, store and process data generated by Security Servers, it's preferred to have only one solution that supports the different data sets.
